### PR TITLE
[Experimental] Add support for FSDP

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: examples/distributed_training.py [hunk 1]
```diff
@@ -87,7 +87,10 @@ def main():
         model=MMResNet50(),
         work_dir='./work_dir',
         train_dataloader=train_dataloader,
-        optim_wrapper=dict(optimizer=dict(type=SGD, lr=0.001, momentum=0.9)),
+        optim_wrapper=dict(
+            optimizer=dict(type=SGD, lr=0.001, momentum=0.9),
+            paramwise_cfg=dict(custom_keys=dict(
+                conv1=dict(lr_mult=0.1)))),
```
Comment (zhouzaida) [inline examples/distributed_training.py:93]: Why do we modify this file?

File: examples/distributed_training_with_flexible_runner.py [hunk 2]
```diff
@@ -43,8 +42,8 @@ def compute_metrics(self, results):
 def parse_args():
     parser = argparse.ArgumentParser(description='Distributed Training')
     parser.add_argument('--local_rank', '--local-rank', type=int, default=0)
-    parser.add_argument('--use-deepspeed', action='store_true')
-
+    parser.add_argument(
+        '--strategy', choices=['deepspeed', 'fsdp', 'ddp'], default='ddp')
```
Comment (zhouzaida) [inline examples/distributed_training_with_flexible_runner.py:46]: This script should also support single device training.

File: mmengine/_strategy/base.py [hunk 3]
```diff
@@ -962,3 +964,13 @@ def collect_env(self) -> Tuple[dict, dict]:
         runtime_env['GPU number'] = self.world_size
 
         return system_env, runtime_env
+
+    def _return_prepared(self):
```
Comment (zhouzaida) [inline mmengine/_strategy/base.py:968]:
```suggestion
    def _prepared_components(self):
```

File: mmengine/_strategy/deepspeed.py [hunk 4]
```diff
@@ -169,12 +171,12 @@ def prepare(
             self.param_schedulers = self.build_param_scheduler(
                 param_scheduler, self.optim_wrapper)
             return_items.append(self.param_schedulers)
-
-        return return_items[0] if len(return_items) == 1 else return_items
+        self._prepared = True
+        return self._return_prepared()
```
Comment (zhouzaida) [inline mmengine/_strategy/deepspeed.py:175]:
```suggestion
        return self._prepared_components()
```

File: examples/distributed_training_with_flexible_runner.py (hunk 2 above)
Comment (HAOCHENYE) [reply examples/distributed_training_with_flexible_runner.py:46]: Set defaults to 'none', which is compatible for single device training and ddp training.

File: mmengine/_strategy/utils.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+            device_idx = [arg.name
+                          for arg in func._schema.arguments].index('device')
+            if len(args) > device_idx:
+                args = list(args)
+                args[device_idx] = 'meta'
+            else:
+                kwargs['device'] = 'meta'
+            return func(*args, **kwargs)
+        else:
+            return func(*args, **kwargs)
```
Comment (zhouzaida) [inline mmengine/_strategy/utils.py:19]:
```suggestion

        return func(*args, **kwargs)
```

File: mmengine/optim/optimizer/amp_optimizer_wrapper.py [hunk 6]
```diff
@@ -48,6 +49,10 @@ class AmpOptimWrapper(OptimWrapper):
             `'float64'`. If set to ``None``, the default data type will be used.
             Defaults to None.
             `New in version 0.6.1.`
+        use_fsdp (bool): Using ``ShardedGradScaler`` when it is True. It should
+            be enabled when using ``FullyShardedDataParallel``
+            Defaults to False.
+            `New in version 0.8.0.`
```
Comment (zhouzaida) [inline mmengine/optim/optimizer/amp_optimizer_wrapper.py:55]:
```suggestion
        use_fsdp (bool): Using ``ShardedGradScaler`` when it is True. It should
            be enabled when using ``FullyShardedDataParallel``.
            Defaults to False.
            `New in version 0.8.0.`
```

File: mmengine/optim/optimizer/amp_optimizer_wrapper.py [hunk 7]
```diff
@@ -73,17 +79,29 @@ def __init__(self,
...
+        if use_fsdp and digit_version(
+                torch.__version__) >= digit_version('2.0.0'):
+            from torch.distributed.fsdp.sharded_grad_scaler import \
+                ShardedGradScaler
+            scaler_type = ShardedGradScaler
+        else:
+            scaler_type = GradScaler
+            if use_fsdp:
+                print_log('Please use PyTorch > 2.0.0 when enable '
+                          '`use_fsdp=True`')
```
Comment (zhouzaida) [inline mmengine/optim/optimizer/amp_optimizer_wrapper.py:92]:
```suggestion
        if use_fsdp
            if digit_version(torch.__version__) >= digit_version('2.0.0'):
                from torch.distributed.fsdp.sharded_grad_scaler import \
                    ShardedGradScaler
                scaler_type = ShardedGradScaler
            else:
                raise RuntimeError('PyTorch>=2.0.0 is required when sets `use_fsdp=True`')
        else:
            scaler_type = GradScaler
```

File: mmengine/_strategy/fsdp.py [hunk 8]
```diff
@@ -0,0 +1,609 @@
...
+                >>>    use_orig_params=True,
+                >>> )
+
+            See more configurable arguments in
+            :class:`MMFullyShardedDataParallel`. Defaults to None
+        skip_init_weights (bool, optional): Whether to skip initialization of
+            weights. Defaults to False. This is useful when the parameters of
+            the large model are loaded from a checkpoint, since skipping the
+            initialization of weights can save a lot of time.
+        state_dict_cfg (Union[str, dict, None], optional): Configuration for
```
Comment (zhouzaida) [inline mmengine/_strategy/fsdp.py:65]:
```suggestion
        state_dict_cfg (str or dict, optional): Configuration for
```

Comment (zhouzaida) [reply mmengine/_strategy/fsdp.py:66]: Can the `state_dict_cfg` be None?

File: mmengine/model/wrappers/fully_sharded_distributed.py [hunk 9]
```diff
@@ -86,80 +98,127 @@ class MMFullyShardedDataParallel(FullyShardedDataParallel):
...
+        if isinstance(auto_wrap_policy, str):
+            auto_wrap_policy = FUNCTIONS.get(  # type: ignore
+                auto_wrap_policy)
+        elif isinstance(auto_wrap_policy, dict):
+            ori_func = FUNCTIONS.get(  # type: ignore
+                auto_wrap_policy.pop('type'))
+            auto_wrap_policy = partial(ori_func, **auto_wrap_policy)
+        if not (callable(auto_wrap_policy)
+                or auto_wrap_policy is None):  # type: ignore
+            raise TypeError('Registered `fsdp_auto_wrap_policy` needs to be '
```
Comment (zhouzaida) [inline mmengine/model/wrappers/fully_sharded_distributed.py:161]:
```suggestion
            raise TypeError('Registered `auto_wrap_policy` needs to be '
```

File: mmengine/model/wrappers/fully_sharded_distributed.py [hunk 10]
```diff
@@ -86,80 +98,127 @@ class MMFullyShardedDataParallel(FullyShardedDataParallel):
...
+                            '`Callable`, but has type '
+                            f'{type(auto_wrap_policy)}')
+
+        if isinstance(backward_prefetch, str):
+            backward_prefetch = BackwardPrefetch[backward_prefetch]
+        if not (isinstance(backward_prefetch, BackwardPrefetch)
+                or backward_prefetch is None):
+            raise TypeError(
+                '`backward_prefetch` should be `None`, string of '
+                '"BACKWARD_PRE" and "BACKWARD_POST",  or '
```
Comment (zhouzaida) [inline mmengine/model/wrappers/fully_sharded_distributed.py:171]:
```suggestion
                '"BACKWARD_PRE" and "BACKWARD_POST", or '
```

File: mmengine/model/wrappers/fully_sharded_distributed.py [hunk 11]
```diff
@@ -196,6 +255,12 @@ def train_step(self, data: dict,
                                 f'list tuple or dict, but got {type(data)}')
         parsed_loss, log_vars = self.module.parse_losses(losses)
         optim_wrapper.update_params(parsed_loss)
+
+        # manually zero_grad fixed parameters,
+        # since they are not in any optimizer
+        # if optim_wrapper.should_update():
+        #     for m in self._fixed_modules:
+        #         m.zero_grad(set_to_none=True)
```
Comment (zhouzaida) [inline mmengine/model/wrappers/fully_sharded_distributed.py:263]: Should these comments be removed?

File: mmengine/_strategy/fsdp.py [hunk 12]
```diff
@@ -0,0 +1,605 @@
...
+                 **kwargs):
+        self._init_state_dict_cfg(state_dict_cfg)
+        model_wrapper = model_wrapper or dict()
+        if not isinstance(skip_init_weights, bool):
+            raise TypeError('skip_init_weights must be a boolean, but got '
+                            f'{type(skip_init_weights)}')
+        self.skip_init_weights = skip_init_weights
+        self.model_wrapper = model_wrapper
+        super().__init__(model_wrapper=model_wrapper, **kwargs)
+        self.last_checkpoint = None
```
Comment (zhouzaida) [inline mmengine/_strategy/fsdp.py]:
```suggestion
        model_wrapper = model_wrapper or dict()
        super().__init__(model_wrapper=model_wrapper, **kwargs)

        self._init_state_dict_cfg(state_dict_cfg)
        if not isinstance(skip_init_weights, bool):
            raise TypeError('skip_init_weights must be a boolean, but got '
                            f'{type(skip_init_weights)}')
        self.skip_init_weights = skip_init_weights
        self.last_checkpoint = None
```

File: tests/test_strategies/test_fsdp.py [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+) -> bool:
+    if recurse:
+        return True  # always recurse
+    return isinstance(module, nn.Linear)
+
+
+@skipIf(not torch.cuda.is_available
+        or digit_version(torch.__version__) < digit_version('2.0.0')
+        or not torch.cuda.is_available(),
+        'Only test FSDP with CUDA and PyTorch >= 2.0.0')
```
Comment (zhouzaida) [inline tests/test_strategies/test_fsdp.py:43]:
```suggestion
@skipIf(digit_version(torch.__version__) < digit_version('2.0.0')
        or not torch.cuda.is_available(),
        'Testing FSDP requires CUDA and PyTorch >= 2.0.0')
```

File: tests/test_strategies/test_fsdp.py [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+        self.assertFalse(strategy.skip_init_weights)
+        strategy = FSDPStrategy(state_dict_cfg='local')
+        self._assert_local(strategy)
+
+        strategy = FSDPStrategy(state_dict_cfg='full')
+        self._assert_full(strategy)
+
+        strategy = FSDPStrategy(
+            state_dict_cfg=dict(
+                state_dict_type=StateDictType.LOCAL_STATE_DICT, ))
```
Comment (zhouzaida) [inline tests/test_strategies/test_fsdp.py:64]:
```suggestion
                state_dict_type=StateDictType.LOCAL_STATE_DICT,))
```

File: mmengine/_strategy/fsdp.py [hunk 15]
```diff
@@ -0,0 +1,604 @@
...
+                # Accelerate initialization by skipping init weights
+                with MetaTensorContext():
+                    model = super().build_model(model)
+                model.to_empty(device='cpu')
+        else:
+            model = super().build_model(model)
+
+        # `id_to_name` will be used to convert the `optim_state_dict` of the
+        # raw optimizer to the `optim_state_dict`
+        #  returned by `FSDP.optim_state_dict` in
```
Comment (zhouzaida) [inline mmengine/_strategy/fsdp.py:167]:
```suggestion
        # returned by `FSDP.optim_state_dict` in
```

File: mmengine/_strategy/fsdp.py [hunk 16]
```diff
@@ -0,0 +1,604 @@
...
+        # save param scheduler state dict
+        if save_param_scheduler and not hasattr(self, 'param_schedulers'):
+            self.logger.warning(
+                '`save_param_scheduler` is True but strategy has no '
+                'param_schedulers attribute, so skip saving parameter '
+                'schedulers')
+            save_param_scheduler = False
+
+        if save_param_scheduler:
+            state_dict['param_schedulers'] = self.scheduler_state_dict()
```
Comment (zhouzaida) [inline mmengine/_strategy/fsdp.py:231]:
```suggestion
        if save_param_scheduler and hasattr(self, 'param_schedulers'):
            state_dict['param_schedulers'] = self.scheduler_state_dict()
```

File: mmengine/_strategy/fsdp.py [hunk 17]
```diff
@@ -0,0 +1,597 @@
...
+        """
+        if is_model_wrapper(model):
+            return
+
+        if self.model_wrapper is None:
+            self.model_wrapper = dict(type='MMFullyShardedDataParallel')
+
+        default_args = dict(
+            module=model, device_id=int(os.environ['LOCAL_RANK']))
+        default_args['type'] = 'MMFullyShardedDataParallel'
```
Comment (zhouzaida) [inline mmengine/_strategy/fsdp.py:129]:
```suggestion
        default_args = dict(
            type='MMFullyShardedDataParallel',
            module=model,
            device_id=int(os.environ['LOCAL_RANK']),
        )
```

File: mmengine/_strategy/fsdp.py [hunk 18]
```diff
@@ -0,0 +1,597 @@
...
+        if not isinstance(self.state_dict_config, StateDictConfig):
+            raise TypeError('state_dict_config must be StateDictConfig, but '
+                            f'got {type(self.state_dict_config)}')
+        if not isinstance(self.optim_state_dict_config, OptimStateDictConfig):
+            raise TypeError('optim_state_dict_config must be '
+                            'OptimStateDictConfig, but got '
+                            f'{type(self.optim_state_dict_config)}')
+
+    def convert_model(self, model: nn.Module) -> nn.Module:
+        return model
```
Comment (zhouzaida) [inline mmengine/_strategy/fsdp.py:426]: remove it

File: mmengine/_strategy/fsdp.py [hunk 19]
```diff
@@ -0,0 +1,597 @@
...
+            raise TypeError('optimizer wrapper should be an OptimWrapper '
+                            f'object or dict, but got {optim_wrapper}')
+
+    def _build_param_scheduler(
+        self,
+        scheduler: Union[_ParamScheduler, Dict, List],
+        optim_wrapper: BaseOptimWrapper,
+        default_args: dict,
+    ) -> List[_ParamScheduler]:
+        """Override this method to Update the scheduler with the reconstructed
```
Comment (zhouzaida) [inline mmengine/_strategy/fsdp.py:547]:
```suggestion
        """Override this method to update the scheduler with the reconstructed
```

File: mmengine/model/wrappers/fully_sharded_distributed.py [hunk 20]
```diff
@@ -219,5 +292,156 @@ def test_step(self, data: dict) -> List[BaseDataElement]:
...
+            fixed_modules = []
+            for sub_module in root_module.children():
+                fixed_modules.extend(
+                    find_fixed_modules_recursively(sub_module))
+            return fixed_modules
+
+        fixed_modules = find_fixed_modules_recursively(module)
+        return fixed_modules
+
+    if digit_version(torch.__version__) <= digit_version('2.0.0'):
```
Comment (zhouzaida) [inline mmengine/model/wrappers/fully_sharded_distributed.py:351]:
```suggestion
    if digit_version(torch.__version__) < digit_version('2.0.1'):
```

**Pull request merged**